larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewTransverseTracksAlgorithm.cc

```diff
@@ -36,19 +55,111 @@ void TwoViewTransverseTracksAlgorithm::CalculateOverlapResult(const Cluster *con
+
+    std::mt19937 randomNumberGenerator(static_cast<std::mt19937::result_type>(
+                pCluster2->GetOrderedCaloHitList().size() + pCluster2->GetOrderedCaloHitList().size()));
+
+    const float pvalue(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromPermutationTest(
+        resampledDiscreteProbabilityVector1, resampledDiscreteProbabilityVector2, randomNumberGenerator, m_nPermutations));
+
+    const float matchingScore(1.f-pvalue);
 
-    if (xOverlap > std::numeric_limits<float>::epsilon())
-        this->GetMatchingControl().GetOverlapMatrix().SetOverlapResult(pCluster1, pCluster2, xOverlap);
+    const float locallyMatchedFraction(CalculateLocalMatchingFraction(resampledDiscreteProbabilityVector1, 
+        resampledDiscreteProbabilityVector2, randomNumberGenerator));
+
+    overlapResult = TwoViewTransverseOverlapResult(matchingScore, resampledDiscreteProbabilityVector1.GetSize(), 
+        correlation, locallyMatchedFraction, twoViewXOverlap);
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float TwoViewTransverseTracksAlgorithm::CalculateLocalMatchingFraction(const DiscreteProbabilityVector &discreteProbabilityVector1,
+    const DiscreteProbabilityVector &discreteProbabilityVector2, std::mt19937 &randomNumberGenerator)
+{
+    if (discreteProbabilityVector1.GetSize() != discreteProbabilityVector2.GetSize() || 
+            0 == discreteProbabilityVector1.GetSize()*discreteProbabilityVector2.GetSize())
+        throw STATUS_CODE_INVALID_PARAMETER;
+
+    if (m_minSamples > discreteProbabilityVector1.GetSize())
+        throw STATUS_CODE_INVALID_PARAMETER;
+
+    std::vector<float> localValues1, localValues2;
+    int nMatchedComparisons(0);
+
+    for (size_t iValue = 0; iValue < discreteProbabilityVector1.GetSize(); ++iValue)
+    {
+        localValues1.emplace_back(discreteProbabilityVector1.GetProbability(iValue));
+        localValues2.emplace_back(discreteProbabilityVector2.GetProbability(iValue));
+        if (localValues1.size() == m_minSamples)
+        {
+            const float localPValue(LArDiscreteProbabilityHelper::CalculateCorrelationCoefficientPValueFromPermutationTest(
+                localValues1, localValues2, randomNumberGenerator, m_nPermutations));
+
+            if ((1.f-localPValue) - m_localMatchingScoreThreshold > std::numeric_limits<float>::epsilon())
+                nMatchedComparisons++;
+            localValues1.erase(localValues1.begin());
+            localValues2.erase(localValues2.begin());
+        }
+    }
+
+    const int nComparisons(static_cast<int>(discreteProbabilityVector1.GetSize())-(static_cast<int>(m_minSamples)-1));
+    if (1 > nComparisons)
+        throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
+
+    return static_cast<float>(nMatchedComparisons)/static_cast<float>(nComparisons);
```

Comment:
Spaces around operator. Might be nice to enclose fraction in brackets.

Reply:
Added nice brackets.